File: app/javascript/embeds/spotify.js
import MediaSite from "media_site"

class Spotify extends MediaSite {
  constructor() {
    super(
      /https:\/\/open\.spotify\.com\/(track|album|playlist|episode|show)\/([a-zA-Z0-9]+)(\?.*)?/,

      async (url) => {
        const match = url.match(this.regex)
        if (match) {
          const type = match[1]
          const id = match[2]
          return `https://open.spotify.com/embed/${type}/${id}`
        }
        return null
      },

      (embedUrl) => {
        const iframe = document.createElement('iframe')
        iframe.src = `${embedUrl}?utm_source=generator&theme=0`
        iframe.width = "100%"
        iframe.height = "152"
        iframe.allow = "autoplay clipboard-write encrypted-media encrypted-media fullscreen picture-in-picture"
        iframe.loading = "lazy"
        iframe.style.borderRadius = "12px"
        return iframe
      }
    )
  }
}

export default Spotify

File: app/models/media_embed/providers/spotify.rb
module MediaEmbed
  module Providers
    class Spotify < Base
      REGEX = %r{\Ahttps://open\.spotify\.com/(track|album|artist|playlist|episode|show)/([a-zA-Z0-9]+)}i

      def render(url)
        match = url.match(REGEX)
        return unless match

        type, id = match.captures
        embed_url = "https://open.spotify.com/embed/#{type}/#{id}?utm_source=generator&theme=0"

        iframe(embed_url,
          width: "100%",
          height: type == "album" ? "450" : "152",
          allow: "autoplay; clipboard-write; encrypted-media; fullscreen; picture-in-picture",
          style: style("border-radius: 12px", "max-width: 660px", "margin-inline: auto"))
      end
    end
  end
end
